### PR TITLE
Add an example showcasing how to implement a custom network behavior and access streams

### PR DESCRIPTION
**Pull request by eserilev**
## Description

#4457 

An example that serves as a copy past-able snippet for users that want to access a stream. Also documents how to implement a custom network behavior. 

The example should:

- Have a binary and a library component
- Binary listens on a random port and prints listen address
- Binary accepts an address to dial and connects to it
- Binary opens a stream, send 32 random bytes, waits for the echo and closes again 

## Notes & open questions

This is currently a draft PR, I will add more details as my work progresses

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by eserilev** [comment]
I'm trying to understand how to implement `Control`.

> A control acts as a "remote" that allows users to request streams without interacting with the `Swarm` directly.

`open_stream` should open a new 'stream'  and return it, without needing to interact with the swarm directly. I'm taking this to mean that we shouldn't be using the Swarm abstraction at all in this example were building. Therefore I have a snippet of code that just creates a TcpStream and returns it. I'm not really sure what to do with the `PeerId` though, how were you envisioning this function should look like?

**Comment by thomaseizinger** [comment]
> > A control acts as a "remote" that allows users to request streams without interacting with the `Swarm` directly.
> 
> `open_stream` should open a new 'stream' and return it, without needing to interact with the swarm directly. I'm taking this to mean that we shouldn't be using the Swarm abstraction at all in this example were building.

We still want to use `Swarm` or to be precise, the `NetworkBehaviour` interface:

- `Behaviour` needs to implement `NetworkBehaviour`
- `Control` should use channels to talk to the `NetworkBehaviour` and / or `ConnectionHandler` for requests such as `open_stream`

Something along the lines of:

- When the user calls `open_stream` on `Control`, we need to send a message into a channel that gets read by the `NetworkBehaviour` within its `poll` function.
- This request then needs to be forwarded to an appropriate `ConnectionHandler`, i.e. one that is connected to a peer with the `PeerId` provided by the user.
	- If we are not connected to this peer, we probably want to return an error through a oneshot back to the future that we returned from `Control::open_stream`
	- If we are connected, we can forward the oneshot to the appropriate `ConnectionHandler` and have it open a stream. The stream can then be sent back to the `Control`.

The main thing we want to achieve is not having users use the `Swarm` APIs as those are _synchronous_ but we want to offer an async/await based interface (without re-implementing all of `rust-libp2p`). Thus, the example you are working on can be seen as a "adapter" that bridges between the async/await world and the synchronous state machine world of rust-libp2p.

Let me know if that makes sense! :)

File: examples/escape-hatch/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    pub fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<(PeerId, Stream)> {
+        todo!()
+    }
+}
+
+impl Behaviour {
+    pub fn new(protocol: StreamProtocol) -> (Self, Control, IncomingStreams) {
+        let (sender, receiver) = mpsc::channel::<StreamMessage>();
```
Comment (thomaseizinger) [inline examples/escape-hatch/src/lib.rs]: The `receiver` needs to go into `Behaviour` so you can call `poll_next` on it and actually _act_ on the message from the `Control`.

File: examples/escape-hatch/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+    receiver: mpsc::Receiver<StreamMessage>
+}
+
+impl IncomingStreams {
+    pub async fn next(&mut self) -> (PeerId, Stream) {
+        if let Some(e) = self.events.pop_back() {
+        } else {
+        }
+
+        todo!()
```
Comment (thomaseizinger) [inline examples/escape-hatch/src/lib.rs:67]: Similar to `Control::open_stream`, this needs to listen on an `mpsc::Receiver` where the `mpsc::Sender` lives in `Behaviour`. Any inbound streams from the `ConnectionHandler` need to be sent this one.

**Comment by thomaseizinger** [comment]
@eserilev Let me know what your appetite is for continuing to work on this. I'll be [stepping down](https://blog.eizinger.io/48326/maintaining-rust-libp2p-beyond) as a maintainer soon and offering this capability is something that would be great to land as a final action. If you are okay with it, I'd be happy to take over from here :)

**Comment by eserilev** [comment]
@thomaseizinger feel free to take it from here, thanks for all the help!!

**Comment by thomaseizinger** [comment]
> @thomaseizinger feel free to take it from here, thanks for all the help!!

Okay, thanks!

**Comment by thomaseizinger** [comment]
Going to close this for now and start my own branch.

**Comment by thomaseizinger** [comment]
Continuing here: https://github.com/libp2p/rust-libp2p/pull/5027.

**Pull request closed without merging**